### PR TITLE
nezuko: surface curvature 4-ep budget-aligned (follow-up #788)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Surface curvature (H̃, K̃) on the surface input path improves surface-side test channels** — confirmed at EP4-partial in PR #788. The previous run was budget-limited at 81% of EP4 with no LR cooldown (run `3ct0x7zd`, `--lr-cosine-t-max 13` schedule). Result: val_abupt 6.7767% (+0.18pp behind SOTA), test_abupt 8.139% (−0.18pp vs within-cluster control thorfinn Arm A). Surface_p, wall_shear, τ_z all beat the within-cluster control on test.

The 270-min train timeout caps us at ~4 effective epochs. Aligning the cosine schedule to the actual budget (`--lr-cosine-t-max 4 --epochs 4`) should let LR fully cool down by end of EP4 and close most of the +0.18pp val_abupt gap to SOTA.

**Predicted outcome:** val_abupt lands in the **6.55–6.75%** range. If <6.5985%, this beats SOTA and merges. If between 6.60–6.80%, the surface-side test signal is real but the budget caps the achievable val gain, in which case the next move is to compose curvature with one of the geometry-OOD interventions (FiLM v3, SDF-FiLM) on a longer dl24-track run.

This is essentially the same architecture as PR #788 with only the schedule changed. The hypothesis is identical; we are testing whether budget-aligned LR cooldown closes the gap.

## Instructions

Re-run the PR #788 surface-curvature configuration with the schedule aligned to the actual 270-min train budget (4 effective epochs).

**Single arm.** No sweep. The previous run's hyperparameters are kept identical except `--lr-cosine-t-max` and `--epochs`, and the kill thresholds are rescaled to the new schedule.

**Reproduce command:**

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surface-curvature-feature \
  --curvature-cache-dir /mnt/new-pvc/Processed/curvatures_haku_v1 \
  --kill-thresholds "10864:val_primary/abupt_axis_mean_rel_l2_pct<32,21728:val_primary/abupt_axis_mean_rel_l2_pct<18,32592:val_primary/abupt_axis_mean_rel_l2_pct<10" \
  --wandb-group nezuko-surface-curvature-4ep \
  --wandb-name nezuko/surface-curvature-4ep
```

**Critical fixes already applied (carry forward from PR #788):**

1. **Kill-threshold operators** must be `<` (PASSING condition is `metric < bound`), NOT `>`. The first PR #788 run (`cspx4pan`) was killed by a healthy 27.19% val_abupt at step 10,864 because the operator was inverted. Triple-check this when copy-pasting.
2. **Curvature cache** is haku PR #580 at `/mnt/new-pvc/Processed/curvatures_haku_v1` (per-surface-point H̃, K̃, signed-log compressed, 484 cases). The edward PR #773 cache stores per-volume-point H,K — wrong tensor shape for surface input.
3. **Curvature normalization** is train-split z-score, NOT raw-unit hard-coded scaling: subtract H mean=−0.0283 (std=0.3298), K mean=+0.1131 (std=0.4569). The cache is already signed-log compressed, so don't double-compress.
4. `surface_input_dim=9` (7 base channels + 2 curvature channels H̃, K̃). Confirm via `project_surface_features.project.weight.shape == [hidden, 102]` at step 0 (= 7 base + 2 curv + 96 STRING-sep RFF, but check the actual numbers against your model's printed shape).

**Vol-points schedule rescaled to 4 epochs:** `0:16384:1:32768:2:49152:3:65536` (was `0:16384:3:32768:6:49152:9:65536` for 13-epoch schedule). This compresses the curriculum so all four point-count bumps actually happen within the 4-epoch budget.

**Kill thresholds rescaled to 4 epochs:**
- step 10864 (end of EP1): `val_primary/abupt_axis_mean_rel_l2_pct<32` (~25–26% expected, healthy at 32%)
- step 21728 (end of EP2): `val_primary/abupt_axis_mean_rel_l2_pct<18` (~7–9% expected with 4-ep cooldown)
- step 32592 (end of EP3): `val_primary/abupt_axis_mean_rel_l2_pct<10` (~6.8–7.2% expected)

Note: rescaled kill thresholds are tighter than 13-epoch run because the 4-epoch cosine cooldown has the LR drop much faster.

## Baseline

**Single-model SOTA target (merge bar):** val_abupt **< 6.5985%** (from PR #592 alphonse `4k25s25e`)

**Within-cluster control:** thorfinn Arm A — test_abupt 8.321%, test_vol_p 12.092%, test_surf_p 4.303%, test_wall_shear 7.697%

**PR #788 reference (previous run, budget-limited at EP4-partial step 37,336):**
- val_abupt 6.7767% (+0.18pp vs SOTA)
- test_abupt 8.139% (−0.18pp vs control ✓)
- test_surface_p 4.168% (−0.14pp vs control ✓)
- test_wall_shear 7.4189% (−0.28pp vs control ✓)
- test_volume_p 12.254% (+0.16pp vs control)
- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/3ct0x7zd

**SOTA reproduce:** `cd target/ && torchrun --standalone --nproc_per_node=8 train.py --agent alphonse --optimizer lion --lr 9e-5 --weight-decay 5e-4 --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 --pos-encoding-mode string_separable --use-qk-norm --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" --lr-cosine-t-max 4 --epochs 4 --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" --no-compile-model --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 --batch-size 4` (PR #592, run `4k25s25e`).

## Reporting

At each EP val checkpoint (step 10864, 21728, 32592, ~43459), post the per-channel breakdown and the EPn→EPn-1 deltas, plus the curvature input statistics (post-z-score mean/std for H̃ and K̃) and the first-layer surface-encoder column-L2 norms for cols 4 (H̃), 5 (K̃) and the original surface extras (cols 0–3) so we can confirm the curvature features are being used at healthy magnitudes (target: 0.1×–2.0× of median(orig extras)).

At completion, post:
1. **Best EMA checkpoint val_primary** (full 34-case): val_abupt, val_surface_p, val_volume_p, val_wall_shear (+ per-axis x/y/z)
2. **Test_surface metrics from best-val checkpoint**: test_abupt, test_surface_p, test_volume_p (rel-l2 % AND MAE), test_wall_shear (+ per-axis)
3. **Three-way comparison** vs SOTA #592 and vs PR #788 (`3ct0x7zd`)
4. **Per-epoch trajectory table** (val_abupt, val_surf_p, val_vol_p, val_wall_shear by EP)
5. **Run health**: peak VRAM, total train minutes, any kill-gate or NaN events

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #795 because SUPERSEDED: PR #798 is the correct schedule-aligned 4-ep surface curvature re-run with the full corrected reproduce command. Closing this duplicate to avoid confusion — please work from PR #798..
